frontend/src/components/mainPage/codeEditor/CodeEditor.vue

```diff
@@ -1,53 +1,264 @@
+            this.updateBreakpointsDecorations(this.breakpointsDecorations);
+            this.updateLineDecorations(this.lineDecorations);
+            this.updateTargetDecorations(this.targetDecorations);
         },
 
-        handlePickVariable(variable) {
+        handlePickVariable(event) {
+            const word = this.editor.getModel().getWordAtPosition(event.target.position);
+            const variable = {
+                name: event.target.element.innerText,
+                start: lineColumn(this.$props.code).toIndex(event.target.position.lineNumber, word.startColumn),
+                end: lineColumn(this.$props.code).toIndex(event.target.position.lineNumber, word.endColumn),
+            };
             this.$emit("pickVariableEvent", variable);
         },
+
+        handleChangeContent(event) {
+            for (let change of event.changes) {
+                // change.forceMoveMarkers is undefined when text is changed by code.
+                // We must update decorations manually, because otherwise it doesn't show up on startup
+                if (change.forceMoveMarkers == undefined) {
+                    this.updateAllDecorations();
+                    continue;
+                }
+
+                let legacyChange = monacoChangeToLegacyFormat(this.$props.code, change);
+                moveTrackedVariables(legacyChange);
+                moveBreakpoints(this.project, legacyChange);
+            }
+        },
+
+        handleClick(event) {
+            if (event.target.element.classList.contains("breakpoint")) {
+                if (!this.$props.editable) return;
+                this.project.breakpoints.addOrDelete({ id: event.target.position.lineNumber - 1 });
+                return;
+            }
+
+            if (event.target.element.classList.contains("target")) {
+                this.handlePickVariable(event);
+                return;
+            }
+        },
+
+        editorDidMount(editor) {
+            this.editor = editor;
+
+            editor.getModel().onDidChangeContent((event) => {
+                this.handleChangeContent(event);
+            });
+
+            editor.onMouseDown((event) => {
+                this.handleClick(event);
+            });
+        },
+
+        updateVariablesDecorations(dec) {
+            this.currentVariablesDecorations = this.editor.deltaDecorations(this.currentVariablesDecorations, dec);
+        },
+
+        updateBreakpointsDecorations(dec) {
+            this.currentBreakpointsDecorations = this.editor.deltaDecorations(this.currentBreakpointsDecorations, dec);
+        },
+
+        updateLineDecorations(dec) {
+            this.currentLineDecorations = this.editor.deltaDecorations(this.currentLineDecorations, dec);
+        },
+
+        updateTargetDecorations(dec) {
+            this.currentTargetDecorations = this.editor.deltaDecorations(this.currentTargetDecorations, dec);
+        },
+
+        getBreakpointClass(i) {
```

Comment:
Ta funkcja chyba bardziej pasuje do "computed", bo tu nie ma wykonywanej żadnej akcji tylko przeliczanie.

Reply:
Ale przyjmuje argument `i`, a w computed nie można przekazywać parametrów.

Reply:
`getBreakpointClass() {
    return (index) => {
        if (this.project.breakpoints.has(index)) return "breakpoint breakpointActive";
        if (!this.project.isRunning && this.$props.editable) return "breakpoint";
        return "";
    }
}`
Da się to zrobić w taki sposób. Ktoś tu chyba nie słuchał na mojej prezentacji o vue xD